src/server/protocol_client.cc

```diff
@@ -56,17 +56,23 @@ static ProtocolClient::SSL_CTX* CreateSslClientCntx() {
   const auto& tls_key_file = GetFlag(FLAGS_tls_key_file);
   unsigned mask = SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
 
-  CHECK_EQ(1, SSL_CTX_use_PrivateKey_file(ctx, tls_key_file.c_str(), SSL_FILETYPE_PEM));
-  const auto& tls_cert_file = GetFlag(FLAGS_tls_cert_file);
+  // Load client certificate if given.
+  if (!tls_key_file.empty()) {
+    CHECK_EQ(1, SSL_CTX_use_PrivateKey_file(ctx, tls_key_file.c_str(), SSL_FILETYPE_PEM));
+    const auto& tls_cert_file = GetFlag(FLAGS_tls_cert_file);
 
-  CHECK_EQ(1, SSL_CTX_use_certificate_chain_file(ctx, tls_cert_file.c_str()));
+    CHECK_EQ(1, SSL_CTX_use_certificate_chain_file(ctx, tls_cert_file.c_str()));
+  }
 
+  // Load custom certificate validation if given.
   const auto& tls_ca_cert_file = GetFlag(FLAGS_tls_ca_cert_file);
   const auto& tls_ca_cert_dir = GetFlag(FLAGS_tls_ca_cert_dir);
 
   const auto* file = tls_ca_cert_file.empty() ? nullptr : tls_ca_cert_file.data();
   const auto* dir = tls_ca_cert_dir.empty() ? nullptr : tls_ca_cert_dir.data();
-  CHECK_EQ(1, SSL_CTX_load_verify_locations(ctx, file, dir));
+  if (file || dir) {
```

Comment:
CA is  now optional as well? Is that right or am I missing something?

Reply:
Yes, it is optional. AFAIU if it is not given openssl will validate the server certificate against the system default which will include let's encrypt etc. But maybe it misses an additional call to do that, I still have to check with an actual certificate.

Reply:
it did miss the additional call :laughing: tested manually to verify.